plugins/thunderbolt/fu-thunderbolt-device.c

```diff
@@ -320,6 +320,75 @@ fu_thunderbolt_device_get_attr_uint16(FuThunderboltDevice *self, const gchar *na
 	return (guint16)val;
 }
 
+static gboolean
+fu_thunderbolt_device_set_port_offline(FuUdevDevice *device)
+{
+	g_autoptr(GError) error_local_offline = NULL;
+	g_autoptr(GError) error_local_rescan = NULL;
+
+	if (!fu_udev_device_write_sysfs(device, "usb4_port1/offline", "1", &error_local_offline)) {
+		g_warning("Setting port offline failed: %s", error_local_offline->message);
+		return FALSE;
+	}
+	if (!fu_udev_device_write_sysfs(device, "usb4_port1/rescan", "1", &error_local_rescan)) {
+		g_warning("Rescan on port failed: %s", error_local_rescan->message);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static void
+fu_thunderbolt_device_set_port_online(FuUdevDevice *device)
+{
+	FuUdevDevice *udev = FU_UDEV_DEVICE(device);
+	g_autoptr(GError) error_local = NULL;
+
+	if (!fu_udev_device_write_sysfs(udev, "usb4_port1/offline", "0", &error_local)) {
+		g_warning("Setting port online failed: %s", error_local->message);
+	}
+}
+
+gboolean
+fu_thunderbolt_device_open(FuDevice *device, GError **error)
+{
+	FuThunderboltDevice *self = FU_THUNDERBOLT_DEVICE(device);
+	GUdevDevice *udev_device = NULL;
+	g_autoptr(GUdevDevice) udev_parent1 = NULL;
+	g_autoptr(GUdevDevice) udev_parent2 = NULL;
+	g_autoptr(FuUdevDevice) parent = NULL;
+	if (self->device_type != FU_THUNDERBOLT_DEVICE_TYPE_RETIMER ||
+	    fu_thunderbolt_device_get_version(self, NULL))
+		return TRUE;
+
+	udev_device = fu_udev_device_get_dev(FU_UDEV_DEVICE(device));
+	udev_parent1 = g_udev_device_get_parent(udev_device);
+	udev_parent2 = g_udev_device_get_parent(udev_parent1);
+	parent = fu_udev_device_new_with_context(fu_device_get_context(FU_DEVICE(self)),
+						 g_steal_pointer(&udev_parent2));
+	return fu_thunderbolt_device_set_port_offline(parent);
+}
+
+gboolean
+fu_thunderbolt_device_close(FuDevice *device, GError **error)
+{
+	FuThunderboltDevice *self = FU_THUNDERBOLT_DEVICE(device);
+	GUdevDevice *udev_device = NULL;
+	g_autoptr(GUdevDevice) udev_parent1 = NULL;
+	g_autoptr(GUdevDevice) udev_parent2 = NULL;
+	g_autoptr(FuUdevDevice) parent = NULL;
+	if (self->device_type != FU_THUNDERBOLT_DEVICE_TYPE_RETIMER)
+		return TRUE;
+
+	udev_device = fu_udev_device_get_dev(FU_UDEV_DEVICE(device));
+	udev_parent1 = g_udev_device_get_parent(udev_device);
+	udev_parent2 = g_udev_device_get_parent(udev_parent1);
+	parent = fu_udev_device_new_with_context(fu_device_get_context(FU_DEVICE(self)),
```

Comment:
since the same thing is used in open and close, I wonder if you should instead make a helper function that can be used in both, or save the information to the device

Reply:
done in https://github.com/fwupd/fwupd/pull/4046/commits/02304c8e6ecd9c8484696bb959507e81b706efb8